File: MetodosJavascript/Algoritmos/V4ParallelBlock.js
/*
Algoritmo tomado del artículo http://paper.ijcsns.org/07_book/201202/20120212.pdf y adaptado con la ayuda de inteligencia artificial

Este algoritmo implementa una versión paralela del método de multiplicación de matrices utilizando bloques y Web Workers.
Divide las matrices de entrada en bloques de tamaño especificado y utiliza Web Workers para realizar la multiplicación de bloques de manera paralela.
Esto puede ayudar a mejorar la eficiencia al aprovechar el procesamiento paralelo disponible en los navegadores modernos.

Integrantes del grupo:
Esteban Julián Ortega Tapie
Jhon Sebastián Zambrano Ariza
David Serna Restrepo
*/

import { runWorker } from './WorkerAlgoritmosParallel/Worker.js';

export async function V4_parallel_block(matrizA, matrizB, matrizResultado, tamanioBloque) {
    // Script que se ejecutará en el Worker
    const workerScript = `
        const { parentPort } = require('worker_threads');

        // Función principal que realiza la multiplicación de matrices en bloques
        parentPort.once('message', (workerData) => {
            const { matrizA, matrizB, matrizResultado, tamanioBloque } = workerData;

            for (var i1 = 0; i1 < matrizResultado.length; i1 += tamanioBloque) {
                for (var j1 = 0; j1 < matrizResultado.length; j1 += tamanioBloque) {
                    for (var k1 = 0; k1 < matrizResultado.length; k1 += tamanioBloque) {
                        for (var i = i1; i < i1 + tamanioBloque && i < matrizResultado.length; i++) {
                            for (var j = j1; j < j1 + tamanioBloque && j < matrizResultado.length; j++) {
                                for (var k = k1; k < k1 + tamanioBloque && k < matrizResultado.length; k++) {
                                    matrizResultado[k][i] += matrizA[k][j] * matrizB[j][i];
                                }
                            }
                        }
                    }
                }
            }

            // Enviar el bloque actualizado de la matriz resultado de vuelta al hilo principal
            parentPort.postMessage({ matrizResultado });
        });
    `;

    // Datos que se enviarán al Worker
    const data = { matrizA, matrizB, matrizResultado, tamanioBloque };

    try {
        // Ejecutar el Worker y esperar el resultado
        const result = await runWorker(workerScript, data);
        // Actualizar la matriz resultado con el resultado del Worker
        matrizResultado = result.matrizResultado;
        return matrizResultado; //Retorna la matriz Resultado
    } catch (error) {
        // Manejo de errores en caso de que el Worker falle
        console.error('Error en el Worker:', error);
    }
}

File: MetodosPython/Algoritmos/V4ParallelBlock.py
'''
Algoritmo tomado del artículo http://paper.ijcsns.org/07_book/201202/20120212.pdf y adaptado con la ayuda de inteligencia artificial

Integrantes del grupo:
Esteban Julián Ortega Tapie
Jhon Sebastián Zambrano Ariza
David Serna Restrepo
'''
"""
    Multiplicación paralela de matrices utilizando bloque de tamaño variable (Versión 4).
    
    Args:
    - matrizA (numpy.ndarray): Matriz de entrada A.
    - matrizB (numpy.ndarray): Matriz de entrada B.
    - matrizResultado (numpy.ndarray): Matriz donde se almacenará el resultado de la multiplicación.
    - tamanioBloque (int): Tamaño del bloque para la división de las matrices.
    
    Returns:
    - numpy.ndarray: Matriz resultado de la multiplicación de A y B.
    """
import numpy as np
from concurrent.futures import ThreadPoolExecutor, as_completed

def V4parallel_block(matrizA, matrizB, matrizResultado, tamanioBloque):
    tamanio = len(matrizResultado)

    def task(i1, j1, k1):
        """
        Función para realizar la multiplicación de matrices en un bloque específico.
        
        Args:
        - i1 (int): Índice inicial de filas.
        - j1 (int): Índice inicial de columnas.
        - k1 (int): Índice inicial de la dimensión de bloque.
        
        Returns:
        - numpy.ndarray: Matriz resultado del bloque específico.
        """
        resultado = np.zeros_like(matrizResultado)# Matriz resultado inicializada con ceros
         # Bucles para recorrer los elementos dentro del bloque
        for i in range(i1, min(i1 + tamanioBloque, tamanio)):
            for j in range(j1, min(j1 + tamanioBloque, tamanio)):
                for k in range(k1, min(k1 + tamanioBloque, tamanio)):
                    resultado[k][i] += matrizA[k][j] * matrizB[j][i] # Multiplicación y acumula en la posición [k][i]
        return resultado # Devuelve el resultado del bloque

    with ThreadPoolExecutor() as executor:
        futures = []
         # Bucle para dividir la multiplicación en bloques y ejecutar en paralelo

        for i1 in range(0, tamanio, tamanioBloque):
            for j1 in range(0, tamanio, tamanioBloque):
                for k1 in range(0, tamanio, tamanioBloque):
                    futures.append(executor.submit(task, i1, j1, k1)) # Envía la tarea al executor y agrega el futuro a la lista

        resultado = np.zeros_like(matrizResultado)# Matriz resultado inicializada con ceros
        for future in as_completed(futures):
            resultado += future.result()# Suma los resultados de los bloques a la matriz resultado final

    return resultado # Devuelve la matriz resultado después de la multiplicación